### ISSUE TITLE
New string functions --- Open for Comments ---

### ISSUE THREAD

**Issue by LiorKogan** [RedisGraph/RedisGraph#2669]
This Issue describes a set of new string functions that we intended to add to RedisGraph:

**--- WORK IN PROGRESS; OPEN FOR COMMENTS ---**

```
- string.removeExcept(`str`, `charGroup` = 1) → string
- string.concat(`list`, `delimiter`) → string
- string.replace(`str`, `regex`, `replacement`) → string
- string.setCase(`str`, `policy`) → string
- string.distance(`str1`, `str2`, `distFunc` = 'Levenshtein') → floating-point
```

**Details:**

---
**string.removeExcept(`str`) → string**

Given a string, return a string after removing all characters except given groups of characters.

If `charGroup` is 1:
  - expect the capitals A..Z

If `charGroup` is 2:
  - expect the small letters a..z

If `charGroup` is 4:
  - expect the digits 0..9
  
If `charGroup` is 8:
  - expect the characters ~!@#$%^&*()[]{}<>,./\|`'"_-+=?

If `charGroup` is 16:
  - expect the space character

`charGroup` can be the sum of the above. For example:

If `charGroup` is 7:
  - Remove any character expect the capitals A..Z, the small letters a..z, and the digits 0..9

Error on:
- Wrong number of arguments (not [1..2])
- `str` is not a string
- `cleanPolicy` is not a valid string
- Memory allocation failure

Examples:
- `list.string.removeExcept('abc ABC 123', 1) → 'ABC'`
- `list.string.removeExcept('abc ABC 123', 7) → 'abcABC123'`

---

**string.concat(`list`, `delimiter` = '') → string**

Given a list of strings and an optional delimiter, return a string which is a concatenation of all the strings with the given delimiter.

Error on:
- Wrong number of arguments (not [1..2])
- `list` is not a list string
- an element of `list` is not a string
- `delimiter` is not a string
- Memory allocation failure

Examples:
- `list.string.concat(['abc', 'def', 'ghi']) → 'abcdefghi'`
- `list.string.concat(['abc', 'def', 'ghi'], '#') → 'abc#def#ghi'`

---
**string.replace(`str`, `regex`, `replacement`) → string**

Given a string, return a string after replacing each regex match with a given replacement.

Error on:
- Wrong number of arguments (not 3)
- `str` is not a string
- `regex` is not a valid string
- `replacement` is not a string
- Memory allocation failure

---

**string.setCase(`str`, `policy`) → string**

Given a string, return a string after changing its case according to a given policy.

If `policy` is 'lower':
  - Replace any A..Z character with a..z

If `policy` is 'upper':
  - Replace any a..z character with A.Z

If `policy` is 'proper': 
  - Capitalize the first letter in each word (any letter that doesn't follow a letter or a digit). Convert all other letters to lower case.
  
If `policy` is 'sentence': 
  - Capitalize the first letter in each sentence (any letter that follows '.' , '!', or '?' followed by whitespaces). Convert all other letters to lower case.

Error on:
- Wrong number of arguments (not 2)
- `str` is not a string
- `policy` is not a valid string
- Memory allocation failure

---

**string.distance(`str1`, `str2`, `distFunc` = 'Levenshtein') → floating-point**

Given two strings, return their distance according to a specified distance function.

If `distFunc` is 'Levenshtein':
  - return the Levenshtein distance

If `distFunc` is 'Hamming':
  - return the Hamming distance

If `distFunc` is 'JaroWinkler':
  - return the Jaro-Winkler distance

Error on:
- Wrong number of arguments (not 3)
- `str1` is not a string
- `str2` is not a string
- `distType` is not a valid string
- Memory allocation failure

**Comment by nafraf** [comment]
Hi, some variables were renamed:
In string.removeExcept(), "Error on" section, ```cleanPolicy``` should be replaced by ```charGroup``` and it's a number
In string.distance(), "Error on" section, ```distType``` should be replaced by ```distFunc```.

Values used in ```distFunc``` and ```policy```, are case insensitive?

How to set the value to Jaro-Winkler scaling factor (p)? or it will be fixed in 0.1?

**Comment by LiorKogan** [comment]
Reference for distance functions: https://maxbachmann.github.io/RapidFuzz/Usage/distance/OSA.html

**Comment by swilly22** [comment]
@LiorKogan

A simplified version of the proposal

`string.concat` should be renamed to `sting.join`

`string.replace` not sure what will it take to support `regex` from the looks of it POSIX provides `regex.h` not sure what's the overhead or what should we do on none POSIX systems

`string.clean` is too complex and can be replaced by `string.replace`

`string.setCase` should be removed, all of that work should be performed by the application side

`string.distance` why do we need this ? do we really need all of different distance functions ?

**Comment by swilly22** [comment]
`string.concat(list, delimiter = '') → string` 👍 
I guess this is similar to python:
```python
','.join(["a", "b", "c"])
```
-----------------------------------------------------------------------------------------------------------------------------------
`string.matchRegEx(str, regex) → array(array(string))` 👍 

-----------------------------------------------------------------------------------------------------------------------------------
`string.replaceRegEx(str, regex, replacement) → string`
consider making `replacement` mandatory, to remove an matches simply set it to the empty string `replacement = ''`

-----------------------------------------------------------------------------------------------------------------------------------
`string.clean(str, exceptChars = 1) → string`

I think there's more flexibility if we allow the user to specify its own an array of `exceptChars` (whitelist) e.g.
`string.clean(str, ['a', 'k'])` remove everything except the 'a' and 'k' chars.

To ease up on the white list generation and add additional flexibility we can extend the `range` function to accept characters e.g. `string.clean(str, range('a', 'z'))`
this will remove everything except lower case 'a' - 'z'.
more over: `string.clean(str, range('A', 'M') + range('a', 'z'))` if you want to go wild.

-----------------------------------------------------------------------------------------------------------------------------------
`string.distance(str1, str2, distFunc = 'Lev', distFuncParams = [...]) → floating-point`

I think it would be easier if `distFuncParams` would be a `map` this way the arguments will be clearly viewed.

**Comment by AviAvni** [comment]
I think string.replaceRegEx is flexible enough so string.clean is not needed string.clean

**Comment by LiorKogan** [comment]
@AviAvni two considerations:
- It can be slower
- We'll need to support "any alphanumeric Unicode character" Regex (see [this](https://stackoverflow.com/questions/5555613/does-w-match-all-alphanumeric-characters-defined-in-the-unicode-standard) and [this](https://www.regular-expressions.info/unicode.html))

**Comment by OfirMos** [comment]
@LiorKogan Currently the library suggested isn't supporting the string.matchRegEx(str, regex) → array(array(string))
functionality, it can only return:
`string.matchRegEx(['blabla <header h1>txt1</header> blabla <header h2>txt2</header>', '<header (\\w+)>(\\w+)</header>') → [["<header h1>txt1</header>"], ["<header h2>txt2</header>"]]`

**Comment by OfirMos** [comment]
@LiorKogan @AviAvni  I found some bugs in Oniguruma library, such handling the following regex: '(?:\\?)' and couple of more similar ones. I couldn't find any other Java compatible library.
I guess I'll be able to fix it, the problem is that there might be many other bugs I haven't found yet.
Google RE2 seems to have a pretty large coverage of the Java regex but it's written in C++ not sure if it's acceptable by @swilly22  
[https://github.com/google/re2/wiki/Syntax](url)

**Comment by swilly22** [comment]
@LiorKogan we've considered `string.clean` and we find the grouping to be a bit excessive, the same functionality can be achieved via `RegEx` replace which I find to be even more flexible.

**Comment by LiorKogan** [comment]
@swilly22 I'm not against it, but make sure we have a valid RegEx expression for `except all letters (multilingual)`

**Comment by LiorKogan** [comment]
@OfirMos yes, it is noncommutative by definition, regardless of the algorithm. `ab->a` requires deletion while `a->ab` requires insertion. I'll add a note to the description.

**Comment by OfirMos** [comment]
@LiorKogan Regarding the Jaro distance, can a character from s2 match twice?
For instance: EFE and EEE

**Comment by OfirMos** [comment]
@LiorKogan I think we are missing threshold parameter with default of 0.7 in the Jaro winkler distance.

**Comment by OfirMos** [comment]
@LiorKogan regarding the Lev: "if InsertionWeight and  if InsertionWeight and SubstitutionWeight are not equal, are not equal,"
Shouldn't it be deletionWeight instead of SubstitutionWeight?

**Comment by LiorKogan** [comment]
@OfirMos yes, fixed.

**Comment by OfirMos** [comment]
@LiorKogan regarding the string distance DLev, for the osa distance can I call it DLev_OSA?

**Comment by LiorKogan** [comment]
Just OSA

**Comment by nafraf** [comment]
@LiorKogan In case of duplicated keys in ``distFuncParams``, the function takes the last value.
Is that right? or the duplicated keys should be reported as an error?

```sh
GRAPH.QUERY g "RETURN string.distance('abc', 'abcd', 'Lev', {InsertionWeight: 0.3, InsertionWeight: 0.7})"
1) 1) "string.distance('abc', 'abcd', 'Lev', {InsertionWeight: 0.3, InsertionWeight: 0.7})"
2) 1) 1) "0.7"
```

**Comment by LiorKogan** [comment]
@nafraf no need to error. Behavior should be consistent with map definition.

**Comment by OfirMos** [comment]
@nafraf You won't be able to error on this, because in python you can't duplicate keys (this is a python dictionary) , so in the module you will only get InsertionWeight one time.
